packages/core/src/components/forms/_control-group.scss

```diff
@@ -76,69 +174,126 @@ Styleguide components.forms.control-group
+    // equally-specific styles.
+    [readonly],
+    :disabled,
+    .pt-disabled {
+      z-index: $control-group-z-index-input-disabled;
+      border-radius: 0;
+    }
 
-  > :first-child {
-    border-radius: $pt-border-radius 0 0 $pt-border-radius;
-  }
+    &[class*="pt-intent"] {
+      z-index: $control-group-z-index-intent-input-default;
 
-  > :last-child {
-    margin-right: 0;
-    border-radius: 0 $pt-border-radius $pt-border-radius 0;
+      .pt-dark & {
+        z-index: $control-group-z-index-dark-theme-intent-input-default;
+      }
+
+      &:focus {
+        z-index: $control-group-z-index-intent-input-focus;
+      }
+    }
   }
 
-  // similarly to button groups, elements in control groups are stacked
-  // in a very particular order for best visual results
   .pt-button,
-  .pt-select {
-    position: relative;
-    z-index: 1;
+  .pt-select select {
+    z-index: $control-group-z-index-button-default;
+    // inherit radius since it's most likely to be zero
+    border-radius: inherit;
 
-    &[class*="pt-intent-"] {
-      z-index: 3;
+    &:focus {
+      // establish new stacking context so focus state covers neighbors
+      position: relative;
+      z-index: $control-group-z-index-button-focus;
     }
-  }
 
-  .pt-input {
-    position: relative;
-    z-index: 0;
+    &:hover {
+      z-index: $control-group-z-index-button-hover;
+    }
 
-    &[class*="pt-intent-"] {
-      z-index: 3;
+    &:active {
+      z-index: $control-group-z-index-button-active;
     }
 
-    &:not([readonly], :disabled, .pt-disabled):focus {
-      z-index: 6;
-      border-radius: $pt-border-radius;
+    &[readonly],
+    &:disabled,
+    &.pt-disabled {
+      z-index: $control-group-z-index-button-disabled;
     }
-  }
 
-  // establish new stacking context so focus state covers neighbors
-  .pt-button:focus,
-  select:focus {
-    position: relative;
-    z-index: 2;
+    &[class*="pt-intent"] {
+      z-index: $control-group-z-index-intent-button-default;
+
+      &:focus {
+        z-index: $control-group-z-index-intent-button-focus;
+      }
+
+      &:hover {
+        z-index: $control-group-z-index-intent-button-hover;
+      }
+
+      &:active {
+        z-index: $control-group-z-index-intent-button-active;
+      }
+
+      &[readonly],
+      &:disabled,
+      &.pt-disabled {
+        z-index: $control-group-z-index-intent-button-disabled;
+      }
+    }
   }
 
   // input group contents appear above input always
   .pt-input-group > .pt-icon,
   .pt-input-group > .pt-button,
-  .pt-input-group > .pt-input-action,
+  .pt-input-group > .pt-input-action {
+    z-index: $control-group-z-index-input-group-children;
+  }
+
+  // keep the select-menu carets on top of everything always
   .pt-select::after {
-    z-index: 7;
+    z-index: $control-group-z-index-select-caret;
+  }
+
+  // have consecutive elements share a border
+  &:not(.pt-vertical) > * {
+    margin-right: -$button-border-width;
```

Comment:
That's actually on true for light theme.

The visuals look different in dark theme, depending on whether you use a button group or single buttons for consecutive buttons (top is bad, bottom is good):
![image](https://cloud.githubusercontent.com/assets/199754/22524348/b52af44c-e877-11e6-8834-7e6371857580.png)

So we can keep that line for the light theme, but we need to take care of dark theme. Proposal (to be refactor in the appropriate place):

```
.pt-dark &:not(.pt-vertical) {
  > * {
    margin-right: 0;
  }

  > .pt-button + .pt-button {
    margin-left: $button-border-width;
  }
}
```